include/range/v3/view/group_by.hpp

```diff
@@ -53,59 +54,49 @@ namespace ranges
     private:
         friend range_access;
         Rng rng_;
+        // cached version of the end of the first subrange
```

Comment:
You probably wouldn't need the comment if `next_cur_` was named like the equivalent of the other views that cache `begin()`.

Reply:
... which would be how? Some of them name it `begin_` (which is wrong for this case) and some of them name it `cache_` (which is meaningless).

Reply:
`begin_` looks right. It _is_ the view's begin.

Reply:
Since `group_by` always includes the first element, `second_` seems like a good name.

These PRs were brought up due to the lack of test coverage. I hope to not have made you choose a solution riddled with more bugs. Fingers crossed that there's an empty range test to make sure `second_`'s not misused.

Reply:
It could be if I made it like:

```cpp
non_propagating_cache<subrange<iterator_t<Rng>>> begin_;
```

That's larger (if we care) and a bit redundant to store `rng_.begin()` twice. But if the cursor has a `subrange<iterator_t<Rng>>` member, that might be clearer (regardless of any other changes).

Reply:
> Fingers crossed that there's an empty range test to make sure second_'s not misused.

Uh... of... uh... Why yes of course there will be an empty range test.

Reply:
Make that tests, for each possible return value of the predicate. You never know.

Reply:
Not to make it sound as it being your fault. I was hoping that they were there since the beginning, and not just one test with an obvious use case. Your answer suggests they weren't ;(

Reply:
Regarding the tests, you could also make use of my [test cases](https://github.com/ericniebler/range-v3/pull/1376/files#diff-d4e7db63ae69a359b671e7c5919644fe), which ensures [the implementation meets the definition](https://github.com/ericniebler/range-v3/pull/1376#issue-346601466). I've tested your changes with them, they pass.

Reply:
Adding your tests, and an empty and singleton test.